### PR TITLE
[en] improve EN_REPEATEDWORDS_NEED

### PR DESCRIPTION
**Pull request by AzadehSafakish**
Added a first test AP for 'need to', as 'require to' is an odd suggestion:

I still need to sign-in somewhere. You need to sign-in too. →
I still need to sign-in somewhere. You **require** to sign-in too.

However, I couldn't get tokenRegex("need(s|ed|ing)?") or even tokenRegex("need|needs|needed|needing") to work, so it currently only works for VBP.

Questions, comments, concerns, tips?
@jaumeortola @udomai

### CODE REVIEW

**Comment by jaumeortola** [comment]
For me, tokenRegex is working. The sentence "It needs to be done. That needs to be done." currently marked as incorrect, becomes correct. Isn't it the expected result?

File: languagetool-language-modules/en/src/test/java/org/languagetool/rules/en/EnglishRepeatedWordsRuleTest.java [hunk 1]
```diff
@@ -49,6 +49,7 @@ public void testRule() throws IOException {
     assertCorrectText("This is a new experience. Happy New Year!");
     assertCorrectText("This was needed. There is a need to do it.");
     assertCorrectText("It needs to be done. That also needed to be done.");
+    assertCorrectText("I still need to sign-in somewhere. You need to sign-in too.");
```
Comment (udomai) [inline languagetool-language-modules/en/src/test/java/org/languagetool/rules/en/EnglishRepeatedWordsRuleTest.java:52]: I think it should be "sign in", just as a side note 🤓

File: languagetool-language-modules/en/src/main/java/org/languagetool/rules/en/EnglishRepeatedWordsRule.java [hunk 2]
```diff
@@ -18,22 +18,44 @@
...
 
 public class EnglishRepeatedWordsRule extends AbstractRepeatedWordsRule{
   
   private static final EnglishSynthesizer synth = new EnglishSynthesizer(new AmericanEnglish());
 
+  private final Supplier<List<DisambiguationPatternRule>> antiPatterns;
+
+  private static final List<List<PatternToken>> ANTI_PATTERNS = Arrays.asList(
+    Arrays.asList(
+      token("need"),   // "I still need -> require to sign-in"
```
Comment (udomai) [inline languagetool-language-modules/en/src/main/java/org/languagetool/rules/en/EnglishRepeatedWordsRule.java:47]: you could make this line

```java
new PatternTokenBuilder().token("need").posRegex("VB.*").matchInflectedForms().build(),
```

but

```java
new PatternTokenBuilder().tokenRegex("need(s|ed|ing)?").build(),
```

should work, too

**Review by udomai — CHANGES_REQUESTED** [review]

File: languagetool-language-modules/en/src/test/java/org/languagetool/rules/en/EnglishRepeatedWordsRuleTest.java (hunk 1 above)
Comment (udomai) [reply languagetool-language-modules/en/src/test/java/org/languagetool/rules/en/EnglishRepeatedWordsRuleTest.java:52]: With one of the solutions in the other comment, this sentence should be correct, too:

```
assertCorrectText("I still needed to sign-in somewhere. She needs to sign-in too.");
```

**Comment by udomai** [comment]
By the way, @jaumeortola, can you confirm the rule matches all forms of "need" (matching the tag and chunk specified in synonyms.txt)? Just making sure.

**Comment by AzadehSafakish** [comment]
> 
> 
> For me, tokenRegex is working. The sentence "It needs to be done. That needs to be done." currently marked as incorrect, becomes correct. Isn't it the expected result?

Yes, that is the expected result. I'm not sure what I'm missing, but when I try it, I believe it no longer catches the error in the test sentences (...)

[ERROR] Failures: 
[ERROR]   EnglishRepeatedWordsRuleTest.testRule:80 expected:<1> but was:<0>

File: languagetool-language-modules/en/src/test/java/org/languagetool/rules/en/EnglishRepeatedWordsRuleTest.java (hunk 1 above)
Comment (AzadehSafakish) [reply languagetool-language-modules/en/src/test/java/org/languagetool/rules/en/EnglishRepeatedWordsRuleTest.java:52]: Perhaps?
I'll drop the hyphen in 'sign-in' once you drop the hyphen in 'extra-careful' XD
What even is that emoji...

**Comment by udomai** [comment]
Line 80 tests the sentence two lines above it: "It needs to be done. That needs to be done." After you insert the AP, this sentence will no longer be caught – i.e., this sentence needs to be changed to something your AP allows.

**Comment by jaumeortola** [comment]
Merged manually.

**Comment by jaumeortola** [comment]
https://github.com/languagetool-org/languagetool/commit/01323a4c23a105841701ddd3758931ad9d6aef0d

**Comment by AzadehSafakish** [comment]
Wait, I see. I was looking in the wrong place.
Thank you so much @jaumeortola !!

**Pull request closed without merging**